File: src/plugins/templates.zig
const std = @import("std");
const plugin = @import("../plugin.zig");
const Editor = @import("../Editor.zig");
const Self = @This();

/// Templates plugin — insert predefined markdown templates.
/// Commands: :tpl.meeting, :tpl.daily, :tpl.readme, :tpl.blog
pub fn pluginInfo(self: *Self) plugin.PluginInfo {
    _ = self;
    return .{ .name = "templates", .version = "0.1.0", .author = "LazyMD contributors", .description = "Markdown document templates" };
}
pub fn pluginInit(self: *Self, editor: *Editor) void {
    _ = self;
    _ = editor;
}
pub fn pluginDeinit(self: *Self) void {
    _ = self;
}
pub fn onEvent(self: *Self, event: *plugin.PluginEvent) void {
    _ = self;
    _ = event;
}
pub fn getCommands(self: *Self) []const plugin.CommandDef {
    _ = self;
    return &[_]plugin.CommandDef{
        .{ .name = "tpl.meeting", .description = "Insert meeting notes template", .handler = &insertMeeting },
        .{ .name = "tpl.daily", .description = "Insert daily note template", .handler = &insertDaily },
        .{ .name = "tpl.readme", .description = "Insert README template", .handler = &insertReadme },
        .{ .name = "tpl.blog", .description = "Insert blog post template", .handler = &insertBlog },
    };
}

fn insertTemplate(editor: *Editor, template: []const u8) void {
    const offset = editor.buffer.posToOffset(editor.cursor_row, editor.cursor_col);
    editor.buffer.insertSlice(offset, template) catch {
        editor.status.set("Failed to insert template", true);
        return;
    };
    editor.status.set("Template inserted", false);
}

fn insertMeeting(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\# Meeting Notes
        \\
        \\**Date:**
        \\**Attendees:**
        \\**Topic:**
        \\
        \\## Agenda
        \\
        \\- [ ] Item 1
        \\- [ ] Item 2
        \\- [ ] Item 3
        \\
        \\## Notes
        \\
        \\
        \\
        \\## Action Items
        \\
        \\- [ ]
        \\
    );
}

fn insertDaily(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\# Daily Note
        \\
        \\## Tasks
        \\
        \\- [ ]
        \\
        \\## Notes
        \\
        \\
        \\
        \\## Journal
        \\
        \\
        \\
    );
}

fn insertReadme(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\# Project Name
        \\
        \\Short description of the project.
        \\
        \\## Installation
        \\
        \\```bash
        \\# install instructions
        \\```
        \\
        \\## Usage
        \\
        \\```bash
        \\# usage example
        \\```
        \\
        \\## Features
        \\
        \\- Feature 1
        \\- Feature 2
        \\
        \\## License
        \\
        \\MIT
        \\
    );
}

fn insertBlog(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\---
        \\title: "Blog Post Title"
        \\date:
        \\tags: []
        \\draft: true
        \\---
        \\
        \\# Blog Post Title
        \\
        \\## Introduction
        \\
        \\
        \\
        \\## Main Content
        \\
        \\
        \\
        \\## Conclusion
        \\
        \\
        \\
    );
}

test "templates plugin info" {
    var p = Self{};
    try std.testing.expectEqualStrings("templates", p.pluginInfo().name);
    const cmds = p.getCommands();
    try std.testing.expectEqual(@as(usize, 4), cmds.len);
}

File: internal/plugins/templates.go
package plugins

import (
	"github.com/EME130/lazymd/internal/editor"
)

type TemplatesPlugin struct{}

func (p *TemplatesPlugin) Info() PluginInfo {
	return PluginInfo{
		Name:        "templates",
		Version:     "0.1.0",
		Author:      "LazyMD contributors",
		Description: "Markdown document templates",
	}
}

func (p *TemplatesPlugin) Init(ed editor.PluginEditor) {}

func (p *TemplatesPlugin) OnEvent(event *PluginEvent) {}

func (p *TemplatesPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "tpl.meeting", Description: "Insert meeting template", Handler: tplMeetingHandler},
		{Name: "tpl.daily", Description: "Insert daily note template", Handler: tplDailyHandler},
		{Name: "tpl.readme", Description: "Insert README template", Handler: tplReadmeHandler},
		{Name: "tpl.blog", Description: "Insert blog post template", Handler: tplBlogHandler},
	}
}

func tplMeetingHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	offset := buf.PosToOffset(ed.CursorRow(), ed.CursorCol())

	template := `# Meeting Notes

**Date:** 2024-01-01
**Attendees:**
**Topic:**

## Agenda

-

## Discussion

-

## Action Items

- [ ]

`
	buf.InsertString(offset, template)
	ed.SetStatus("Meeting template inserted", false)
}

func tplDailyHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	offset := buf.PosToOffset(ed.CursorRow(), ed.CursorCol())

	template := `# Daily Note

## Tasks

- [ ]

## Notes

`
	buf.InsertString(offset, template)
	ed.SetStatus("Daily note template inserted", false)
}

func tplReadmeHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	offset := buf.PosToOffset(ed.CursorRow(), ed.CursorCol())

	template := `# Project Title

## Description

Brief description of the project.

## Installation

` + "```bash" + `
# Installation commands
` + "```" + `

## Usage

` + "```bash" + `
# Usage examples
` + "```" + `

## Contributing

Contributions are welcome!

## License

MIT

`
	buf.InsertString(offset, template)
	ed.SetStatus("README template inserted", false)
}

func tplBlogHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	offset := buf.PosToOffset(ed.CursorRow(), ed.CursorCol())

	template := `---
title: Post Title
date: 2024-01-01
tags: []
draft: true
---

# Post Title

Introduction paragraph.

## Section 1

Content here.

## Conclusion

Closing thoughts.

`
	buf.InsertString(offset, template)
	ed.SetStatus("Blog post template inserted", false)
}
